### PR TITLE
Use IDL-based COM interop generator

### PR DESCRIPTION
**Pull request by kekekeks**

### CODE REVIEW

File: src/tools/MicroComGenerator/CppGen.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        
+        public static string GenerateCpp(AstIdlNode idl)
+        {
+            var sb = new StringBuilder();
+            var preamble = idl.GetAttributeOrDefault("cpp-preamble");
+            if (preamble != null)
+                sb.AppendLine(preamble);
+
+            foreach (var s in idl.Structs)
+                sb.AppendLine("struct " + s.Name + ";");
```
Comment (KvanTTT) [inline src/tools/MicroComGenerator/CppGen.cs:28]: Split? Here and below.

```suggestion
            foreach (var s in idl.Structs)
            {
                sb.AppendLine("struct ");
                sb.AppendLine(s.Name);
                sb.AppendLine(";");
            }
```

Comment (kekekeks) [reply src/tools/MicroComGenerator/CppGen.cs:28]: Well, I'm not really worried about efficiency of CPP codegen *that much* (it takes 6ms on my machine), StringBuilder is here mostly for readability. The Roslyn one, on the other hand, takes whole 2 seconds.

Comment (kekekeks) [reply src/tools/MicroComGenerator/CppGen.cs:28]: image

File: src/Avalonia.Build.Tasks/ComInteropHelper.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                var instructions = method.Body.Instructions;
+                for (int i = 0; i < instructions.Count; i++)
+                {
+                    Instruction instruction = instructions[i];
+
+                    if (instruction.OpCode == OpCodes.Call && instruction.Operand is MethodReference)
+                    {
+                        var methodDescription = (MethodReference)instruction.Operand;
+
+                        if (methodDescription.Name.StartsWith("Calli") && methodDescription.DeclaringType.Name == "LocalInterop")
```
Comment (jkoritzinsky) [inline src/Avalonia.Build.Tasks/ComInteropHelper.cs:72]: I'd recommend using function pointers and C# module initializers once we build with the .NET 5 SDK, but this patching works for now.

**Review by jkoritzinsky — APPROVED** [review]
This looks great!

There's a few improvements we could do if we wanted to preserve object identity (1 CCW/RCW per object/pointer), but that's not necessary for Avalonia to my knowledge.

**Pull request merged**